enhancements/management-workload-partitioning.md

```diff
@@ -0,0 +1,549 @@
+the normal scheduling criteria could be used to safely place them on
+CPUs in the shared pool. This proposal reframes the problem, so that
+instead of considering "scaling" or "tuning" the requests we think about
+"isolating" or "partitioning" the workloads away from the non-management
+workloads. This view of the problem is more consistent with how the
+requirement was originally presented by the customer.
+
+We want to define "management workloads" in a flexible way. For the
+purposes of this document, "management workloads" include all
+OpenShift core components necessary to run the cluster, any add-on
+operators necessary to make up the "platform" as defined by telco
+customers, and operators or other components from third-party vendors
+that the customer deems as management rather than operational. It is
+important to note that not all of those components will be delivered
+as part of the OpenShift payload and some may be written by the
+customer or by vendors who are not our partners.
+
+The basic proposal is to provide a way to identify management
+workloads at runtime and to use CRI-O to run them on a user-selected
+set of CPUs, while other workloads will be prevented from running
+there. This effectively gives 3 pools of CPUs (shared, dedicated, and
+management) and means the shared CPU pool will need other CPUs to
+support burstable or best-effort workloads.
+
+We want the isolation to be in place from the initial buildout of the
+cluster, to ensure predictable behavior. Therefore, the feature must
+be enabled during installation. To enable the feature, the user will
+specify the set of CPUs to run all management workloads. When the
+management CPU set is not defined, the feature will be completely
+disabled.
+
+We generally want components to opt-in to being considered management
+workloads. Therefore, for a regular pod to be considered to contain a
+management workload it must be labeled with `io.openshift.management:
+true`.
+
+We want to treat all OpenShift components as management workloads,
+including those that run the control plane. Therefore, kubelet will be
+modified to treat all static pods as management workloads, when the
+feature is enabled. We will update operator manifests and
+implementations to label all OpenShift components not running in
+static pods with the `io.openshift.management: true` label, and add a
+CI job to require that label be present in all workloads created from
+the release payload.
+
+We need kubelet to know when the feature is enabled, but we cannot
+change the configuration schema that kubelet gets from
+upstream. Therefore we will have kubelet look for a new configuration
+file on startup, and the feature will only be enabled if that file is
+found.
+
+We want to give cluster administrators control over which workloads
+are run on the management CPUs. Therefore, only pods in namespaces
+labeled with `io.openshift.management: true` will be subject to
```

Comment:
Namespaces aren't sufficient.  We will have namespaces that have mixed management/not-management workloads and it's not appropriate for us to be forced to split those into different namespaces.

The RBAC behavior can be addressed simply by requiring high/admin privilege to set this on a pod, possibly by the same component that does the cpu request -> workload transition.

Reply:
cc: @deads2k

Reply:
Is it possible to say that someone needs a specific role in order to add a specific label to a pod?

What namespaces do we have with mixed workloads?

Reply:
i agree.
openshift monitoring is a good example.
it has per-node components and per infra host components (prometheus).
in standalone deployments with workers, we only want to compress per-node components.

Reply:
I'm not sure that what you're saying is incompatible with what is being proposed.

The idea with the pod label is to opt-in to being put on the management pool. The namespace label is about security, and giving the cluster admin control over what gets to use management pools at all.

Are we saying that we can't use the namespace because that gives anyone in that namespace the ability to run on the special pool?

Reply:
we already do perms checks in pod admission, this is no different.

Reply:
The text changed enough that this comment thread is detached from the diff view, but I don't think we can resolve it until we hear from @smarterclayton